File: tauri-app/src/types/Schedule.ts
export interface Schedule {
  id?: string;
  scheduleId?: string;
  description?: string;
  items: ScheduleItem[];
}

// Base fields common to all schedule items
export interface BaseScheduleItem {
  itemId: string;
  description: string;
  isRecording: boolean;
  startTime?: number;
  endTime?: number;
}

// Media Items
export interface AudioMediaItem extends BaseScheduleItem {
  itemType: 'audio';
  url: string;
  typeId: string;
}

export interface VideoMediaItem extends BaseScheduleItem {
  itemType: 'video';
  url: string;
  typeId: string;
}

export interface ImageMediaItem extends BaseScheduleItem {
  itemType: 'image';
  url: string;
  typeId: string;
}

export interface TextContentItem extends BaseScheduleItem {
  itemType: 'text-content';
  url: string;
  typeId?: string;
}

export interface YleAudioMediaItem extends BaseScheduleItem {
  itemType: 'yle-audio';
  url: string; // YLE program ID
}

export interface YleVideoMediaItem extends BaseScheduleItem {
  itemType: 'yle-video';
  url: string; // YLE program ID
}

// Prompt Items
export interface ChoicePromptItem extends BaseScheduleItem {
  itemType: 'choice';
  options: string[];
}

export interface MultiChoicePromptItem extends BaseScheduleItem {
  itemType: 'multi-choice';
  options: string[];
  otherEntryLabel?: string;
}

export interface SuperChoicePromptItem extends BaseScheduleItem {
  itemType: 'super-choice';
  options: string[];
  otherEntryLabel?: string;
}

export interface TextInputItem extends BaseScheduleItem {
  itemType: 'text-input';
}

// Discriminated union of all schedule item types
export type ScheduleItem =
  | AudioMediaItem
  | VideoMediaItem
  | ImageMediaItem
  | TextContentItem
  | YleAudioMediaItem
  | YleVideoMediaItem
  | ChoicePromptItem
  | MultiChoicePromptItem
  | SuperChoicePromptItem
  | TextInputItem;

// Schedule Item State
export interface ScheduleItemState {
  title?: Record<string, string>;
  body1?: Record<string, string>;
  body2?: Record<string, string>;
  imageUrl?: string;
}

export enum ScheduleItemStateType {
  Start = 'start',
  Recording = 'recording',
  Finish = 'finish',
}

// Type guard helpers
export function isMediaItem(item: ScheduleItem): item is AudioMediaItem | VideoMediaItem | ImageMediaItem | TextContentItem | YleAudioMediaItem | YleVideoMediaItem {
  return ['audio', 'video', 'image', 'text-content', 'yle-audio', 'yle-video'].includes(item.itemType);
}

export function isPromptItem(item: ScheduleItem): item is ChoicePromptItem | MultiChoicePromptItem | SuperChoicePromptItem | TextInputItem {
  return ['choice', 'multi-choice', 'super-choice', 'text-input'].includes(item.itemType);
}

File: tauri-app/src-tauri/src/models/schedule.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Schedule {
    pub id: Option<String>,
    pub schedule_id: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub items: Vec<ScheduleItem>,
}

/// Schedule item with polymorphic variants using inline structs
/// The itemType field acts as the discriminator for JSON deserialization
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "itemType", rename_all = "kebab-case")]
pub enum ScheduleItem {
    // Media items
    Audio {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String,
        #[serde(rename = "typeId")]
        type_id: String,
    },
    Video {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String,
        #[serde(rename = "typeId")]
        type_id: String,
    },
    Image {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String,
        #[serde(rename = "typeId")]
        type_id: String,
    },
    #[serde(rename = "text-content")]
    TextContent {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String,
        #[serde(rename = "typeId")]
        type_id: Option<String>,
    },
    #[serde(rename = "yle-audio")]
    YleAudio {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String, // YLE program ID
    },
    #[serde(rename = "yle-video")]
    YleVideo {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        url: String, // YLE program ID
    },
    
    // Prompt items
    Choice {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        options: Vec<String>,
    },
    #[serde(rename = "multi-choice")]
    MultiChoice {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        options: Vec<String>,
        #[serde(rename = "otherEntryLabel")]
        other_entry_label: Option<String>,
    },
    #[serde(rename = "super-choice")]
    SuperChoice {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
        options: Vec<String>,
        #[serde(rename = "otherEntryLabel")]
        other_entry_label: Option<String>,
    },
    #[serde(rename = "text-input")]
    TextInput {
        #[serde(rename = "itemId")]
        item_id: String,
        description: String,
        #[serde(rename = "isRecording")]
        is_recording: bool,
        #[serde(default, rename = "startTime")]
        start_time: i32,
        #[serde(default, rename = "endTime")]
        end_time: i32,
    },
}
